File: scripts/infinite_sequence.py
class InfiniteSequence:
    """
    A utility for creating infinite sequences with safe iteration limits
    to prevent runaway loops in client code.
    """
    def __init__(self, start=0, step=1):
        self.start = start
        self.step = step
        
    def generate(self):
        """Generator that yields sequence values infinitely."""
        current = self.start
        while True:
            yield current
            current += self.step

    def safe_take(self, limit=100):
        """
        Safely takes up to `limit` values from the infinite sequence,
        ensuring the iteration terminates.
        """
        gen = self.generate()
        result = []
        for _ in range(limit):
            result.append(next(gen))
        return result

if __name__ == '__main__':
    seq = InfiniteSequence(start=1, step=2)
    print("Generating first 10 odd numbers safely:")
    print(seq.safe_take(10))

File: apps/api/src/infinite-sequence.ts
/**
 * Infinite sequence generator with safe iteration
 */
export function* infiniteSequence(start: number = 0, step: number = 1): Generator<number> {
  let current = start;
  while (true) {
    yield current;
    current += step;
  }
}

export function take<T>(generator: Generator<T>, count: number): T[] {
  const result: T[] = [];
  for (let i = 0; i < count; i++) {
    result.push(generator.next().value);
  }
  return result;
}

export function* fibonacci(): Generator<number> {
  let a = 0, b = 1;
  while (true) {
    yield a;
    [a, b] = [b, a + b];
  }
}

export function* primes(): Generator<number> {
  yield 2;
  let n = 3;
  while (true) {
    if (isPrime(n)) yield n;
    n += 2;
  }
}

function isPrime(n: number): boolean {
  if (n < 2) return false;
  if (n === 2) return true;
  if (n % 2 === 0) return false;
  for (let i = 3; i <= Math.sqrt(n); i += 2) {
    if (n % i === 0) return false;
  }
  return true;
}
